go-controller/pkg/ovn/base_network_controller_secondary.go

```diff
@@ -839,12 +839,38 @@ func (bsnc *BaseSecondaryNetworkController) buildUDNEgressSNAT(localPodSubnets [
 		if masqIP == nil {
 			return nil, fmt.Errorf("masquerade IP cannot be empty network %s (%d): %v", bsnc.GetNetworkName(), networkID, err)
 		}
-		snats = append(snats, libovsdbops.BuildSNATWithMatch(&masqIP.ManagementPort.IP, localPodSubnet, outputPort,
-			extIDs, getMasqueradeManagementIPSNATMatch(dstMac.String())))
+		if !isUDNAdvertised {
+			snats = append(snats, libovsdbops.BuildSNATWithMatch(&masqIP.ManagementPort.IP, localPodSubnet, outputPort,
+				extIDs, getMasqueradeManagementIPSNATMatch(dstMac.String())))
+		} else {
+			dbIDs := getEgressIPAddrSetDbIDs(NodeIPAddrSetName, types.DefaultNetworkName, DefaultNetworkControllerName)
+			addrSet, err := bsnc.addressSetFactory.GetAddressSet(dbIDs)
+			if err != nil {
+				return nil, fmt.Errorf("cannot ensure that addressSet %s exists %v", NodeIPAddrSetName, err)
+			}
+			ipv4ClusterNodeIPAS, ipv6ClusterNodeIPAS := addrSet.GetASHashNames()
+			snats = append(snats, libovsdbops.BuildSNATWithMatch(&masqIP.ManagementPort.IP, localPodSubnet, outputPort,
+				extIDs, fmt.Sprintf("%s && (%s)", getMasqueradeManagementIPSNATMatch(dstMac.String()), getClusterNodesDestinationBasedSNATMatch(ipv4ClusterNodeIPAS, ipv6ClusterNodeIPAS))))
+		}
 	}
 	return snats, nil
 }
 
+func getMasqueradeManagementIPSNATMatch(dstMac string) string {
+	return fmt.Sprintf("eth.dst == %s", dstMac)
+}
+
+func getClusterNodesDestinationBasedSNATMatch(ipv4ClusterNodeIPAS, ipv6ClusterNodeIPAS string) string {
+	if config.IPv4Mode && config.IPv6Mode {
+		return fmt.Sprintf("ip4.dst == $%s || ip6.dst == $%s", ipv4ClusterNodeIPAS, ipv6ClusterNodeIPAS)
+	} else if config.IPv4Mode {
+		return fmt.Sprintf("ip4.dst == $%s", ipv4ClusterNodeIPAS)
+	} else if config.IPv6Mode {
+		return fmt.Sprintf("ip6.dst == $%s", ipv6ClusterNodeIPAS)
+	}
+	return ""
```

Comment:
would this happen ? if it happens, snat match would become invalid.

Reply:
Empty match means that the match isn't effective, it will be regular SNAT, which I'm not sure if that's the desire there